[hotfix] Check for external auth when claiming an unregistered user [PLAT-1070]

erinspace opened the pull request:
If a user is authenticated with external auth, confirm the claim when clicking the link sent by a contributor instead of requiring an OSF password. This is because a user with external auth does not have an OSF password to enter, and we'll trust the external auth to be valid.

<!-- Before submit your Pull Request, make sure you picked
     the right branch:

     - For hotfixes, select "master" as the target branch
     - For new features, select "develop" as the target branch
     - For release feature fixes, select the relevant release branch (release/X.Y.Z) as the target branch -->

## Purpose

A user who has signed in to the OSF through an external service (right now only ORCID) cannot then claim an unregistered user via a link sent to them by a project's contributor, because this relies on a password page. An externally verified user has no password.

## Changes

- An externally-verified user no longer is asked for a password to confirm once they've clicked the link in their email
- Test for an ORCID verified user clicking a password confirm link from an email

<!-- Briefly describe or list your changes  -->

## QA Notes

Steps to reproduce:
1. Create a new OSF account via ORCID.
2. Attempt to claim an unregistered contributor on a node.
3. Have the node creator forward you the claim link.
4. Click the claim link
5. You should now be a contributor on the project

You can also repeat the above but log out of the ORCID user before clicking the claim link. Once you log back in via ORCID, you should be confirmed as a contributor.

## Documentation

<!-- Does any internal or external documentation need to be updated?
     - Developer documentation? If so, link developer.osf.io PR here.
-->
None needed

## Side Effects

none anticipated
<!-- Any possible side effects? -->

## Ticket
https://openscience.atlassian.net/browse/PLAT-1070

cslzchen commented on website/project/views/contributor.py:
```diff
@@ -635,7 +643,7 @@ def claim_user_registered(auth, node, **kwargs):
     current_user = auth.user
 
-    sign_out_url = web_url_for('auth_register', logout=True, next=request.url)
+    sign_out_url = cas.get_login_url(service_url=request.url)
```
It just occur to me that `get_logout_url` is always necessary since it is possible that OSF is logged out but CAS is logged in ...

erinspace commented:
Thank you so much for your review @cslzchen! Yes, testing it out locally with an actual ORCID login would be amazing, I was only able to do so locally with a fudged ORCID-only login user.

I did a bit of looking and yes, it does seem like the logout param is only used by `login_and_register_handler` and `auth_register` for the purposes of `claim_user_registered`. I went ahead and removed it, partially to see if tests passed OK, but also to see what we thought of maybe including this change, or maybe it's too big for a hotfix?

cslzchen commented:
I agree. How about leave the `auth_register()` and `login_and_register_handler()` intact for now?Maybe simply add a TODO comment with a new low/medium priority ticket? It would take a great amount of extra time and effort to test and verify changes on both if they were updated. In this way, we can still deploy this as a quick hot-fix.

erinspace commented:
Ok! Went ahead and reverted those changes, added some TODOs in their place with a reference to [a new ticket](https://openscience.atlassian.net/browse/PLAT-1151) for a more thorough look into removing `logout` from those methods. 

Thanks again @cslzchen!

Johnetordoff commented on website/project/views/contributor.py:
```diff
@@ -665,22 +672,26 @@ def claim_user_registered(auth, node, **kwargs):
-                    trust=False
-                )
-                return redirect(node.url)
+                should_claim = True
```
I think with this you want to put the exit condition first, it's a bit cleaner like this:
```python
    # If a user is already validated though external auth, it is OK to claim
    if check_external_auth(auth.user):
        node.replace_contributor(old=unreg_user, new=current_user)
        node.save()
        status.push_status_message(
            'You are now a contributor to this project.',
            kind='success',
            trust=False
        )
        return redirect(node.url)

    form = PasswordForm(request.form)
    if request.method == 'POST':
        if not form.validate():
            forms.push_errors_to_status(form.errors)
        if not current_user.check_password(form.password.data):
            status.push_status_message(language.LOGIN_FAILED, kind='warning', trust=False)
```

erinspace replied:
If the user isn't externally authenticated they should still go through the existing check on the form - it's only if they're *only* authenticated through ORCID that they get to skip the form entirely. This way seems to assume that external auth is the only path to authentication, when really it's either being externally validated or having an OSF login and still needing to re-enter your password. Repeating the claim auth code in both of those situations seemed redundant hence the variable check.
